### DISCUSSION TITLE
Executing a sibling effect in the handler

### DISCUSSION THREAD

**Discussion by MonadicT** [haskell-effectful/effectful#167] [General]
I have some Haskell code that I am rewriting from mtl to use effectful and have an issue when the handler needs to execute another effect. I've cooked up some bogus code to illustrate my problem. I added `Foo :> es` to the context of `runFoo` to keep GHC happy about `i <- foo` and ended up with the code as below.
 
```
data Foo :: Effect where
  Foo :: Foo m Int
  Bar :: Foo m String

type instance DispatchOf Foo = Dynamic

foo = send Foo

bar = send Bar

runFoo ::
  Foo :> es =>
  Eff (Foo : es) a ->
  Eff es a
runFoo = interpret $ \env -> \case
  Foo -> pure 1
  Bar -> do
    i <- foo
    pure "Bar"

```

However, `runEff .  runFoo  $ foo` complains that there is no Foo in context. I have tried tinkering with `interpose` in `runFoo` handler, but I don't understand the library well enough to figure this out. Help would be much appreciated.

**Comment by arybczak** [comment]
Your `runFoo` signature says "in order to run the Foo effect you need the Foo effect". Something like 

```
runFoo ::
  Blah :> es =>
  Eff (Foo : es) a ->
  Eff es a
```

would make more sense, it would be then "in order to run the Foo effect you need the Blah effect" and you could use Blah inside the handler.

**Reply by MonadicT** [reply]
Yes, that makes sense if I want to run Blah from Foo. But, my intent is to be able to run the effect Foo from Foo. To add a bit more context, I have a monad with actions that use other actions defined in the same monad. My goal is to rewrite the monad-based implementation with one that is effects-based and not just write a thin wrapper. A similar approach to effects is reasonable and should be possible, no?

**Reply by arybczak** [reply]
Aha, I see.

```
runFoo ::
  Eff (Foo : es) a ->
  Eff es a
runFoo = interpret $ \env -> \case
  Foo -> pure 1
  Bar -> do
    i <- localSeqUnlift env $ \unlift -> unlift foo
    pure "Bar"
```

This should do what you want. By entering the local environment you gain access to the handler of the effect you're currently defining.

**Reply by MonadicT** [reply]
Thanks a lot! It's exactly what I need.
